src/main/java/org/truffleruby/extra/ffi/Pointer.java

```diff
@@ -0,0 +1,187 @@
+        writeBytes(offset, bytes, start, length);
+        writeByte(offset + length, (byte) 0);
+    }
+
+    @CompilerDirectives.TruffleBoundary
+    public void writeString(long offset, String string, int maxLength, Charset cs) {
+        ByteBuffer buf = cs.encode(string);
+        int len = Math.min(maxLength, buf.remaining());
+        writeZeroTerminatedBytes(offset, buf.array(), buf.arrayOffset() + buf.position(), len);
+    }
+
+    @CompilerDirectives.TruffleBoundary
+    public void writeBytes(long offset, long size, byte value) {
+        for (long n = 0; n < size; n++) {
+            writeByte(offset + n, value);
+        }
+    }
+
+    @CompilerDirectives.TruffleBoundary
+    public void writeBytes(long offset, byte[] bytes, int index, int length) {
+        for (int n = 0; n < length; n++) {
+            writeByte(offset + n, bytes[index + n]);
+        }
+    }
+
+    public byte readByte(long offset) {
+        return UNSAFE.getByte(address + offset);
+    }
+
+    @CompilerDirectives.TruffleBoundary
+    public void readBytes(long from, byte[] buffer, int bufferPos, int i) {
+        for (int n = 0; n < i; n++) {
+            buffer[bufferPos + n] = readByte(from + n);
+        }
+    }
+
+    public short readShort(long offset) {
+        return UNSAFE.getShort(address + offset);
+    }
+
+    public int readInt(long offset) {
+        return UNSAFE.getInt(address + offset);
+    }
+
+    public long readLong(long offset) {
+        return UNSAFE.getLong(address + offset);
+    }
+
+    private byte[] readZeroTerminatedByteArray(long offset) {
+        final int length = findByte(offset, (byte) 0);
+        final byte[] bytes = new byte[length];
+        readBytes(offset, bytes, 0, length);
+        return bytes;
+    }
+
+    @CompilerDirectives.TruffleBoundary
+    public String readString(long offset) {
+        return Charset.defaultCharset().decode(ByteBuffer.wrap(readZeroTerminatedByteArray(offset))).toString();
+    }
+
+    @CompilerDirectives.TruffleBoundary
+    public Pointer readPointer(long offset) {
+        final long p = readLong(offset);
+        if (p == 0) {
+            return null;
+        } else {
+            return new Pointer(p);
+        }
+    }
+
+    @CompilerDirectives.TruffleBoundary
+    public int findByte(long offset, byte value) {
+        int n = 0;
+        while (true) {
+            if (readByte(offset + n) == value) {
+                return n;
+            }
+            n++;
+        }
+    }
+
+    public void free() {
+        UNSAFE.freeMemory(address);
+    }
+
+    public long getAddress() {
+        return address;
+    }
+
+    public jnr.ffi.Pointer getPointer() {
+        return Runtime.getSystemRuntime().getMemoryManager().newPointer(address);
+    }
+
+    @CompilerDirectives.TruffleBoundary
+    public void enableAutorelease(FinalizationService finalizationService) {
+        if (autorelease) {
+            return;
+        }
```

Comment:
This is not thread-safe if two thread call enableAutorelease().
The easiest would be to synchronize this method.
Alternatively, we could CAS with a AtomicIntegerFieldUpdater.